### PR TITLE
Notebook for the text generation tutorial

### PR DESCRIPTION
**Pull request by mari-linhares**
Notebook for the text generation tutorial with eager mode.

### CODE REVIEW

**Comment by mari-linhares** [comment]
I signed it!

**Comment by lamberta** [comment]
Oh, nice, thanks!
Including the colab link for review: http://colab.research.google.com/github/mari-linhares/docs/blob/patch-1/site/en/tutorials/sequences/text_generation.ipynb

**Comment by wolffg** [comment]
Hey, this is great; thanks for submitting this.

I'm still looking into this, but be aware we have another Shakespeare generator example for TPUs: 
https://colab.sandbox.google.com/github/tensorflow/tpu/blob/master/tools/colab/shakespeare_with_tpu_and_keras.ipynb

They are different samples.  The TPU one uses numpy input and calls model.fit(); this runs eagerly using gradient tape and uses datasets.  It might be useful to update the title of this one to say "with an RNN in eager TensorFlow" or something like that.

**Comment by mari-linhares** [comment]
Hey @wolffg,

Thanks for letting me know about the TPU example, definitely agree about changing the title.

**Comment by mari-linhares** [comment]
Hey Mark, same! Hope everything is well!

> Nothing major. LMK what you think.

I agree with the changes and already merged them.

To be honest I didn't focus on the code, instead directed most of the effort in adding text and diagrams. So thank you for cleaning up and improving code quality!

**Comment by mari-linhares** [comment]
I'll add a note that this was tested on TensorFlow version 1.11.0 (Tested locally in 1.10.1 and it doesn't seem to work), makes sense? The problem on 1.10.1 is:

```
ValueError: You tried to call `count_params` on embedding, but the layer isn't built. You can build it manually via: `embedding.build(batch_input_shape)`.
```

**Comment by MarkDaoust** [comment]
> I'll add a note that this was tested on TensorFlow version 1.11.0 

SGTM

> ValueError: You tried to call `count_params` on embedding, but the layer isn't built. You can build it manually via: `embedding.build(batch_input_shape)`.

There's now an explicit `build` just before the `model.summary`... so I don't understand why it would say this.

**Comment by lamberta** [comment]
Hey Marianne, do you have the original file for the diagrams? I think we need to make a change now for `max_length` to `seq_length`

**Comment by mari-linhares** [comment]
I do, I'll share the drawings with you (@lamberta), @MarkDaoust and @random-forest.

**Comment by mari-linhares** [comment]
>  I think we need to make a change now for max_length to seq_length
Done.

File: site/en/tutorials/sequences/text_generation.ipynb [hunk 1]
```diff
@@ -819,14 +819,14 @@
     "* The hidden state returned by the model is fed back into the model so that it now has more context, instead than only one word. After predicting the next word, the modified hidden states are again fed back into the model, which is how it learns as it gets more context from the previously predicted words.\n",
     "\n",
     "\n",
-    "![](https://drive.google.com/uc?export=view&id=1MeU-bnClsvknBWTCCPgbrCyGRoyaZf9k)\n",
+    "![](https://drive.google.com/uc?export=view&id=1ZdZmGCfti_QhBrmg06teYhO2-zKl-fMV)\n",
```
Comment (lamberta) [inline site/en/tutorials/sequences/text_generation.ipynb]: Can you add these images to the git repo?
Just create an images directory in the collection, ex: /tutorials/sequences/images/

Comment (mari-linhares) [reply site/en/tutorials/sequences/text_generation.ipynb]: Done! Thank you!

**Comment by yashk2810** [comment]
I got rid of tf.multinomial in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/generative_examples/text_generation.ipynb. Using tf.argmax is easy to understand according to me 😀. Also there is a pending PR to make some changes in the prediction loop to incorporate multi word input. I don't know if that PR has been merged or not. But the change is to do tf.argmax(predictions[-1]) so that it will take the last word to predict the next word in case of multi word input

**Comment by MarkDaoust** [comment]
Hi Yash, 

Thanks for taking a look.

> I got rid of tf.multinomial in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/generative_examples/text_generation.ipynb. 

I prefer the `multinomial` version. It keeps things interesting, and ensures that you get a different output if you reuse a prompt. The `argmax` version is equivalent to temperature~=0.

> Also there is a pending PR to make some changes in the prediction loop to incorporate multi word input. 

I don't quite understand... can you find the PR?

Is anyone planning to stub out https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/generative_examples/text_generation.ipynb and point all the button-links to this file?

**Comment by yashk2810** [comment]
tf.multinomial is also fine. The PR that I was talking about is here (https://github.com/tensorflow/tensorflow/pull/22247#issuecomment-421064307).

**Comment by MarkDaoust** [comment]
Thanks Yash... I'll merge those changes (or something equivalent) in here.

**Comment by yashk2810** [comment]
Also regarding the hidden states, I think having the states getting returned shows the flexibility that tensorflow offers (just what I think :smile:). The stateful=True parameter makes me uncomfortable :rofl:

**Comment by lamberta** [comment]
> Is anyone planning to stub out https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/generative_examples/text_generation.ipynb and point all the button-links to this file?

Please.

Offtopic, slightly, but Mark, can you do the Git subtree trick to bring those notebooks over here without losing history?

File: site/en/tutorials/sequences/text_generation.ipynb [hunk 2]
```diff
@@ -861,7 +861,7 @@
     "\n",
     "    # using a multinomial distribution to predict the word returned by the model\n",
     "    predictions = predictions / temperature\n",
-    "    predicted_id = tf.multinomial(predictions, num_samples=1)[0][0].numpy()\n",
+    "    predicted_id = tf.multinomial(predictions[:-1], num_samples=1)[0][0].numpy()\n",
```
Comment (yashk2810) [inline site/en/tutorials/sequences/text_generation.ipynb]: Shouldn't it be 
tf.multinomial(predictions, num_samples=1)[-1][0].numpy() ?

**Comment by MarkDaoust** [comment]
@yashk2810 

Here's an example of why `multinomial` helps. My first try at a prediction run with `temperature=1e-6`:

```
QUEEN MARGARET:
Why, then he stands between two clergymen!

BUCKINGHAM:
Why, then he stands between two clergymen!

BUCKINGHAM:
Why, then he stands between two clergymen!

BUCKINGHAM:
Why, then he stands between two clergymen!

BUCKINGHAM:
Why, then he stands between two clergymen!

...
```
The randomness keeps you out of loops like this.

**Comment by MarkDaoust** [comment]
> `stateful=True` parameter makes me uncomfortable.

OOP. The objects hold their state.  We're not doing anything with that state, there's no point of brininging it out of the object.

File: site/en/tutorials/sequences/text_generation.ipynb (hunk 2 above)
Comment (MarkDaoust) [reply site/en/tutorials/sequences/text_generation.ipynb]: This is a good point... something is funny right here.

The model is violating the [batch, sequence, depth] shape convention.

I'll fix it give me a minute.

**Comment by yashk2810** [comment]
Ah I see, tf.multinomial sounds good to me :)

**Comment by lamberta** [comment]
Please do! :)

**Comment by MarkDaoust** [comment]
This should be resolvable automatically.

[this comment](https://github.com/tensorflow/docs/pull/94#issuecomment-426744916) has instructions for rebasing.

But keep a copy of your branch in case it blows up.

**Comment by lamberta** [comment]
That's one way. It'd be nice to keep everything here, can you make a pull request to @mari-linhares's branch? https://github.com/mari-linhares/docs/tree/patch-1

**Comment by yashk2810** [comment]
let me try that

**Comment by lamberta** [comment]
Right, you can't push directly to it, but you can submit a pull request and she can accept it (just like this!)

**Comment by yashk2810** [comment]
Done. https://github.com/mari-linhares/docs/pull/2/

Stupid question: Can you guys directly push to her branch because you are owners? @lamberta

**Comment by mari-linhares** [comment]
About the conflict with the text_ganeration notebook, I'm kind of lost about the changes that should be kept or not, I thought it was because of the patch from @yashk2810, that's why I reversed (can you make the PR again, with a more descriptive name? Sorry about this).

Can someone resolve the conflicts? Thank you!

**Comment by MarkDaoust** [comment]
Well that didn't work. This is duplicating a bunch of history.. let's try something else.

Sorry about the mess.

https://github.com/tensorflow/docs/pull/104

I don't understand why I can't push directly to the PR branch (or I would have done it).
If you want to fix the history in the PR here the commands would be:

```
git fetch https://github.com/markdaoust/docs pr94
git branch -f patch-1
git push -f origin patch-1
```

Or we can just merge #104.

**Comment by yashk2810** [comment]
The best way I can think of is creating a new PR to change the original notebook with the changes done here and removing this PR 😅. There might be a better way but the git force is not with me 😂

**Comment by MarkDaoust** [comment]
Done. This is merged by #104.

Thanks Again.

**Comment by lamberta** [comment]
Ok, https://github.com/tensorflow/docs/pull/104 has been merged with all the history.
Thanks, everyone. And thank you Marianne for this pull request.

**Comment by MarkDaoust** [comment]
https://www.tensorflow.org/tutorials/sequences/text_generation

**Pull request closed without merging**